src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -1765,10 +1778,25 @@ internal void EnsureCompilationEventQueueCompleted()
             }
         }
 
+        internal void RegisterPossibleUpcomingEventEnqueue()
+        {
+            Interlocked.Increment(ref _eventQueueEnqueuePendingCount);
+        }
+
+        internal void UnregisterPossibleUpcomingEventEnqueue()
+        {
+            Interlocked.Decrement(ref _eventQueueEnqueuePendingCount);
+        }
+
         internal void CompleteCompilationEventQueue_NoLock()
         {
             RoslynDebug.Assert(EventQueue != null);
 
+            if (Volatile.Read(ref _eventQueueEnqueuePendingCount) != 0)
+            {
+                SpinWait.SpinUntil(() => Volatile.Read(ref _eventQueueEnqueuePendingCount) == 0);
```

Comment:
nit: is the initial if check needed?

Reply:
Concretely it seems to avoid a lambda allocation. More generally though it I usually see the quick test then `SpinUntil` pattern when I see this in the wild. Can see a few examples here 

https://github.com/dotnet/runtime/blob/c3d1d753b3444871d42b6856d7a748ad33225d01/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/TaskCompletionSource.cs#L135

Overall though if we did decide to remove this `if` we should not do so in this PR. This code is being ported as is to minimize risk for VS servicing.